src/graphics/reproject-texture.js

```diff
@@ -281,74 +306,94 @@ const generateGGXSamples = (numSamples, specularPower, sourceTotalPixels) => {
 };
 
 // pack float samples data into an rgba8 texture
-const packSamplesTex = (device, name, samples) => {
-    const numSamples = samples.length;
-
-    const w = Math.min(numSamples, 512);
-    const h = Math.ceil(numSamples / w);
-    const data = new Uint8Array(w * h * 4);
-
-    // normalize float data and pack into rgba8
-    let off = 0;
-    for (let i = 0; i < numSamples; ++i) {
-        packFloat32ToRGBA8(samples[i * 4 + 0] * 0.5 + 0.5, data, off + 0);
-        packFloat32ToRGBA8(samples[i * 4 + 1] * 0.5 + 0.5, data, off + 4);
-        packFloat32ToRGBA8(samples[i * 4 + 2] * 0.5 + 0.5, data, off + 8);
-        packFloat32ToRGBA8(samples[i * 4 + 3] / 8, data, off + 12);
-        off += 16;
-    }
-
+const createSamplesTex = (device, name, samples) => {
+    const packedSamples = packSamples(samples);
     return new Texture(device, {
         name: name,
-        width: w,
-        height: h,
+        width: packedSamples.width,
+        height: packedSamples.height,
         mipmaps: false,
         minFilter: FILTER_NEAREST,
         magFilter: FILTER_NEAREST,
-        levels: [data]
+        levels: [packedSamples.data]
     });
 };
 
-// cache of float sample data packed into rgba8 textures
-const samplesTexCache = { };
+// simple cache storing key->value
+// missFunc is called if the key is not present
+class SimpleCache {
+    constructor() {
+        this.map = new Map();
+    }
 
-const generateLambertSamplesTex = (device, numSamples, sourceTotalPixels) => {
-    const key = `lambert-samples-${numSamples}-${sourceTotalPixels}`;
+    get(key, missFunc) {
+        if (!this.map.has(key)) {
+            const result = missFunc();
+            this.map.set(key, result);
+            return result;
+        }
+        return this.map.get(key);
+    }
+
+    clear() {
+        this.map.clear();
+    }
+}
+
+// per-device cache
+class DeviceCache {
+    constructor() {
+        this.cache = new SimpleCache();
+    }
 
-    if (samplesTexCache.hasOwnProperty(key)) {
-        return samplesTexCache[key];
+    get(device, key, missFunc) {
+        return this.cache.get(device, () => {
+            return new SimpleCache();
+        }).get(key, () => {
+            return missFunc();
+        });
+    }
+
+    clear() {
+        this.cache.clear();
     }
+}
+
+// cache of samples. we store these seperately from textures since multiple devices can use the same
+// set of samples.
```

Comment:
I wonder if the complication is worth the benefit here? If I understand correctly, this cache saves raw data, and the texture cache saves the same data stored in the texture format per device? Maybe we only need the texture cache, and if there's a miss, then regenerate raw data? It's optimising the uncommon case of multiple devices.

Reply:
Generating the data isn't free and the additional complexity is pretty minimal, maybe 4 lines of code?

The original issue being addressed will benefit from caching the data too, so I think it's probably worth it.

Reply:
fair enough